eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/LedgerSpec.hs

```diff
@@ -166,3 +184,19 @@ spec = do
     submitTx_ $
       mkBasicTx $
         mkBasicTxBody & withdrawalsTxBodyL .~ Withdrawals [(ra, mempty)]
+
+  describe "Mempool events" $ do
+    it "No Mempool events should be emitted via LEDGERS rules " $ do
+      nes <- use impNESL
+      slotNo <- use impLastTickG
+      let ls = nes ^. nesEsL . esLStateL
+          pp = nes ^. nesEsL . curPParamsEpochStateL
+          account = nes ^. nesEsL . esAccountStateL
+      tx <- fixupTx $ mkBasicTx mkBasicTxBody
+      Right (_, evs) <-
+        tryRunImpRule @"LEDGERS"
+          (LedgersEnv slotNo pp account)
+          ls
+          (Seq.singleton tx)
+      let mempoolEvents = [ev | LedgerEvent ev@(MempoolEvent (ConwayMempoolEvent _)) <- evs]
+      mempoolEvents `shouldBeExpr` []
```

Comment:
I  tried  to write this test: 
```
    it "Mempool events should be emitted via applyTxs" $ do
      let tx = mkBasicTx mkBasicTxBody
      globals <- use impGlobalsL
      slotNo <- use impLastTickG
      nes <- use impNESL

      (evs, _) <- listen $ applyTxs globals slotNo (Seq.singleton tx) nes
 ```

But in order to use `applyTxs`, it appears that I need a `MonadError (ApplyTxError era) (ImpTestM era)` instance .

I tried to write it, but failed and gave up! 
Tried something like this:
```
  (Era era, Show (PredicateFailure (EraRule "LEDGER" era))) =>
     MonadError (ApplyTxError era) (ImpTestM era)
  where
  throwError (ApplyTxError predFailures) = fail (show predFailures)

  catchError (ImpTestM action) handler =  action  --  I don't know how to thread in the predFailures from the handler   
```

I don't know whether all of this is worth it, or whether it's all misguided and there is some straightforward way to call `applyTxs` within ImpTestM.   I'm curious!  (and a little sad that I couldn't write the instance implementation :crying_cat_face:  )

Reply:
What you need is use `runExcept` in order to get `applyTx` out of `MonadError`

Reply:
Or even simpler just restrict it to either:
```haskell
case applyTxs globals slotNo (Seq.singleton tx) nes of
  Left err -> error $ show err
  Right (st, vTx) -> ...
```

Reply:
I've created a ticket for this: https://github.com/IntersectMBO/cardano-ledger/issues/4651